Diagnostic HUDs

GunHog opened the pull request:
Adds the Diagnostic HUD and NV Diagnostic HUD!
- HUD item that shows health and status of Cyborgs, AI, and Mechs.
- Shows Cyborg health, Alive/Dead/KO status, and cell charge.
- Shows AI status and health.
- Shows mech health, cell charge, and internal damage, if any.
- Integrated to AI to show health of borgs and mechs a Malf would have interest in stealing.
- Integrated to borgs, most useful to Engineering borgs to for repair triage.
- Integrated into RD's hardsuit helmet.

![screenshot 2015-10-28 00 38 39](https://cloud.githubusercontent.com/assets/4955510/10780524/472918c0-7d0c-11e5-9d6b-7db482dad2d7.png)
#### Known Issues

~~\- HUDs do not show on newly spawned entities until HUD is taken off and activated again.~~
- Sanity check required to prevent runtime - cause unknown.
- Code may need improvement - Passive HUD code is awful to navigate and modify properly.
- ~~Sprites may need revision. I do not remember who gave them for me for this, however.~~

~~The RD's hardsuit currently has one integrated, but will likely be removed in the final version due to a bug where wearing a diagnostic HUD item with helmet on will cause the HUD to be lost if either the helmet or HUD item is removed.~~
#### Sprites by 420weedscopes and @WJohn!

Commit:
Diag HUDs 2

Adds Diagnostic HUDs

Commit:
Adds Mechs and Borg cells to HUD

- Cyborg cells are now tracked.
- Mech health, cells, and status now tracked

Commit:
Adds HUD items

- Adds Diagnostic HUD to Sci
- Adds NV Diagnostic HUD as well

RemieRichards commented on code/modules/mob/living/silicon/silicon.dm:
```diff
@@ -25,9 +27,15 @@
+/mob/living/silicon/New()
+	var/datum/atom_hud/data/diagnostic/diag_hud = huds[DATA_HUD_DIAGNOSTIC]
+	diag_hud.add_to_hud(src)
+	..()
```
..() is called first before any changes in the override, usually.
Just consistency y'know.

GunHog replied:
Will fix.

WJohn commented:
:+1: With the sprites, even if they're just orange huds.

Commit:
HUD fixes

- Data HUDs that apply only to humans (medical, security) are now
prepared only for humans, preventing errors.
- The RD's hardsuit helmet now deactivates its own HUD if a diagnostic
HUD item is worn.
- HUDs on newly spawned entities now update properly.

Commit:
Merge remote-tracking branch 'remotes/upstream/master' into Diag_Huds

Conflicts:
	code/modules/mob/living/silicon/robot/robot.dm

GunHog commented:
I cannot into mapping, so maybe I should add two to the robotics wardrobe locker?

Metacide commented:
Adding them to the locker sounds good, and they look pretty handy. If you don't add them to the lockers and this gets merged, can you @ me so I remember to add them to MetaStation with the next update?

:+1:

tkdrg commented:
This fixes #5863

GunHog commented:
No, it does not. I removed the airlock and machine stuffs. It only works on borgs, AI, and mechs currently. I will add bots eventually once the simple animal thing is added.

Adding airlock status things is even more of a major pain to add properly, and people were stating balance issues with it regarding antags shocking doors.

Cookiesndip commented:
Very handy for any antag subverting the AI. +

Razharas commented:
@GunHog 
Tell me when its done

GunHog commented:
Alrighty! I really do need some sprites and to improve the code a little. I could probably squeeze a few out of WJohn for Xeno code!

Commit:
Sprites and fixes

- New sprites by WJohn! Yay!
- Fixes borg cells not updating

GunHog commented:
New Health Bar sprites and cell power bar! Thanks WJ!

GunHog commented:
Well, I have not found any *obvious* problems yet. Review when you like, @Razharas!

Razharas commented:
Didnt notice any if(ckey == "GunHog") so its good enough for me, fix bugs when they are reported

phil235 commented on code/modules/clothing/spacesuits/hardsuit.dm:
```diff
@@ -318,16 +318,26 @@
+/obj/item/clothing/head/helmet/space/hardsuit/rd/dropped(mob/living/carbon/human/user)
 	..(user)
+	user.scanner.devices -= 1
+	if(onboard_hud_enabled && !(user.glasses && istype(user.glasses, /obj/item/clothing/glasses/hud/diagnostic)))
```
this is a hacky fix, and it doesn't even solve the same issue when the items are switched (removing your diag glasses while wearing the helmet).
